### PR TITLE
Optimization: convert `ResolveRequest` into readonly struct

### PR DESCRIPTION
**Pull request by SergeiPavlov**
This will save GC work (one allocation per resolving).
Also:
* No need for null-checking
* Remove `GetOrCreateInstanceThrowsArgumentNullExceptionWhenResolveRequestIsNull` test as not-applicable anymore
* Add `in` function argument specifier to pass pointer instead of struct contents.

### CODE REVIEW

**Comment by alistairjevans** [comment]
Hi @SergeiPavlov, could you please run a few of our benchmarks (I'm thinking DeepGraphResolve and ChildScopeResolve at least) and show before and after numbers here to prove that the benefit of allocation removal outweighs the one or two places we copy `ResolveRequest` into a class field in terms of the size of those classes and the copy time?

If you're unsure on how to run our benchmarks (under the 'bench' folder), let me know.

**Comment by SergeiPavlov** [comment]
Sure. Here are the results:

Branch develop:

ChildScopeResolve

|                          Method |     Mean |    Error |   StdDev |   Gen0 |   Gen1 | Allocated |
|-------------------------------- |---------:|---------:|---------:|-------:|-------:|----------:|
|                         Resolve | 26.78 us | 0.277 us | 0.272 us | 3.7231 | 1.8311 |  30.51 KB |
| ResolveNeverRegisteredFromChild | 13.15 us | 0.067 us | 0.053 us | 2.0905 | 1.0376 |   17.1 KB |


DeepGraphResolve

|  Method |     Mean |     Error |    StdDev |   Gen0 |   Gen1 | Allocated |                                                                                              |-------- |---------:|----------:|----------:|-------:|-------:|----------:|                                                                                              | Resolve | 9.775 us | 0.0253 us | 0.0249 us | 1.7700 | 0.0458 |  14.54 KB |                                                                                                                                                                                 |  Method |     Mean |     Error |    StdDev |   Gen0 |   Gen1 | Allocated |                                                                                              |-------- |---------:|----------:|----------:|-------:|-------:|----------:|                                                                                              | Resolve | 9.775 us | 0.0253 us | 0.0249 us | 1.7700 | 0.0458 |  14.54 KB |                                                                                                                                                                                 |  Method |     Mean |     Error |    StdDev |   Gen0 |   Gen1 | Allocated |
|-------- |---------:|----------:|----------:|-------:|-------:|----------:|
| Resolve | 9.775 us | 0.0253 us | 0.0249 us | 1.7700 | 0.0458 |  14.54 KB |


====================================

Branch  struct_ResolveRequest (THIS)

ChildScopeResolve

|                          Method |     Mean |    Error |   StdDev |   Gen0 |   Gen1 | Allocated |
|-------------------------------- |---------:|---------:|---------:|-------:|-------:|----------:|
|                         Resolve | 26.97 us | 0.270 us | 0.341 us | 3.6621 | 1.8005 |  30.13 KB |
| ResolveNeverRegisteredFromChild | 12.87 us | 0.033 us | 0.029 us | 2.0905 | 1.0376 |  17.12 KB |


DeepGraphResolve

|  Method |     Mean |     Error |    StdDev |   Gen0 |   Gen1 | Allocated |
|-------- |---------:|----------:|----------:|-------:|-------:|----------:|
| Resolve | 9.438 us | 0.1794 us | 0.2395 us | 1.6937 | 0.0305 |  13.85 KB |

**Review by alistairjevans — COMMENTED** [review]
I have no issue with the changes here, but adding `in` is a binary-breaking change, as is changing `ResolveRequest` to a struct; which really should make this a major version bump.

Sure, most people will use the extension methods, but I almost guarantee someone somewhere has their own extension method that instantiates `ResolveRequest` and calls `ResolveComponent`.

We may have to do a release for the .NET 8 release soonish, so I'm tempted to hold this change until we're prepared to do a major version bump.

File: src/Autofac/ResolveRequest.cs [hunk 1]
```diff
@@ -56,4 +56,33 @@ public ResolveRequest(Service service, ServiceRegistration serviceRegistration,
...
+    /// </summary>
+    /// <param name="left">The left operand.</param>
+    /// <param name="right">The right operand.</param>
+    /// <returns>The result of the operator.</returns>
+    public static bool operator !=(ResolveRequest left, ResolveRequest right) =>
+        !(left == right);
+
+    /// <inheritdoc/>
+    public override int GetHashCode() =>
+        Service.GetHashCode() ^ Registration.GetHashCode() ^ ResolvePipeline.GetHashCode() ^ Parameters.GetHashCode() ^ (DecoratorTarget?.GetHashCode() ?? 0);
```
Comment (botinko) [inline src/Autofac/ResolveRequest.cs:87]: Could HashCode.Combine be used here?

Comment (SergeiPavlov) [reply src/Autofac/ResolveRequest.cs:87]: No, because this project  is compilable also for `.NET Standard`

**Comment by tillig** [comment]
Related: autofac/Autofac.Extensions.DependencyInjection#112

I'm not sure we can support M.E.DI keyed services without core Autofac changes. If we have to cut a new/breaking version for .NET 8, it might be worth trying to get these keyed service updates in place at the same time or we could end up with two breaking changes in a row.

**Pull request merged**